File: HR vacancies (hh.ru)/hr_vacancies.R
# Библиотеки ----
library(httr) #для работы с HTTP
library(urltools) #для работы с URL
library(jsonlite) #JSON парсер
library(purrr) #для применения семйства map()
library(furrr) #параллельные вычисления
library(tidyverse) #вселенная инструментов tidyverse
library(RPostgres) #коннектор PostgreSQL




# Основной код ----
# Cоздаем справочники #####

## Профессиональные области HR
prof_id <- c(17,  #бизнес-тренер
             38,  #директор по персоналу
             69,  #менеджер по персоналу. Больше 2000 !
             117, #специалист по кадрам. Больше 2000 !
             118, #специалист по подбору персонала. Больше 2000 !
             153, #менеджер по компенсациям и льготам
             171) #руководитель отдела персонала


## Регионы
areas_url <- "https://api.hh.ru/areas" #путь к API hh.ru с регионами и городами
areas <- content(GET(areas_url), as = "parsed")  #парсим содержимое
areas_id <- map_chr(areas[[1]]$areas, "id") #извлекаем список id российских регионов

## Опыт работы
exp_id <- c('noExperience', #нет опыта
            'between1And3', #от 1 до 3 лет
            'between3And6', #от 3 до 6 лет
            'moreThan6')    #более 6 лет



# Определяем функции #####

#Функция для сбора вакансий в промежуточные датафреймы
get_vacancies_inter <- function(vacancies){
  
  data.frame(
    
    id = vacancies$items$id, #id вакансии
    
    name = vacancies$items$name, #название вакансии
    
    area_id = vacancies$items$area$id, #id региона
    area_name = vacancies$items$area$name, #название города
    
    professional_roles_id = sapply(vacancies$items$professional_roles, "[[", 1), #извлекакам id проф. роли
    professional_roles_name = sapply(vacancies$items$professional_roles, "[[", 2), #извлекакам название проф. роли. Используем функцию sapply(), так это данные вложены в лист
    
    employer_id = vacancies$items$employer$id, #id работодателя
    employer_name = vacancies$items$employer$name, #название компании работодателя
    snippet_requirement = vacancies$items$snippet$requirement, #требования к кандидату
    snippet_responsibility = vacancies$items$snippet$responsibility, #описание обязанностей
    
    experience = vacancies$items$experience$name, #требуемый опыт
    employment = vacancies$items$employment$name, #тип занятости
    
    
    salary_from = ifelse("salary" %in% names(vacancies$items) && "from" %in% names(vacancies$items$salary), vacancies$items$salary$from, NA), #нижняя граница оплаты труда,
    salary_to = ifelse("salary" %in% names(vacancies$items) && "to" %in% names(vacancies$items$salary), vacancies$items$salary$to, NA), #верхняя граница оплаты труда
    salary_currency = ifelse("salary" %in% names(vacancies$items) && "currency" %in% names(vacancies$items$salary), vacancies$items$salary$currency, NA), #валюта
    salary_gross =  ifelse("salary" %in% names(vacancies$items) && "gross" %in% names(vacancies$items$salary), vacancies$items$salary$gross, NA), #признак того, что зарплата указана до вычета налогов.
    #Зарпалата указана не у всех вакансий, поэтому, чтобы не получить ошибку мы используем конструкцию ЕСЛИ ТО, которая проверяет наличие данных о зп.
    
    created_at = vacancies$items$created_at, #дата создания вакансии
    published_at = vacancies$items$published_at, #дата публикации вакансии
    
    url = vacancies$items$alternate_url, #ссылка на вакансию на сайте hh.ru
    
    stringsAsFactors = FALSE #НЕ превращать текстовые поля в факторы
    
  )
  
}



#Итоговая функция по сбору вакансий в один датафрейм

get_vacancies_result <- function(page, prof_id, area_id = NULL, exp_id = NULL) {
  
  #Реализуем логику, отраженную на блок-схеме
  if (prof_id %in% c(17, 38, 153, 171)) {
    query_list = list(professional_role = prof_id, per_page = 100, page = page)
  } else if (area_id == 1) {
    query_list = list(professional_role = prof_id, area = area_id, experience = exp_id, per_page = 100, page = page)
  } else {
    query_list = list(professional_role = prof_id, area = area_id, per_page = 100, page = page)
  }
  
  #Парсим результат по заданному условию, используем для аутентификации OAuth токен
  response <- GET(url = vacancies_url, 
                  add_headers(Authorization = paste("Bearer", access_token)), 
                  query = query_list
                  
                  #если бы вы хотели осуществлять поиск по тексту в названии вакансии
                  # query =  list(text = "hr аналитик", search_field = "name", per_page = 100, page = page))
                  
                  )
  
  #Извлекаем данные из JSON в лист
  vacancies <- fromJSON(rawToChar(response$content))
  
  #Чтобы избежать ошибки, на тот случай, когда страниц менее 20, пишем такую проверку
  if (length(vacancies$items) == 0) {
    return(NULL)
  }
  
  #Соединям промежуточные датафреймы в один итоговый
  #Делаем это с интревалом в 0.5 сек, чтобы не получить бан со стороны API
  vacancies_df_inter <- get_vacancies_inter(vacancies)
  Sys.sleep(0.5)
  
  #Возвращаем результат
  return(vacancies_df_inter)
  
}


# Собираем вакансии #####
access_token <- 'XXXX' #OAuth токен

vacancies_url <- 'https://api.hh.ru/vacancies' #путь к вакансиям API hh.ru

## Определяем сетки параметров
### Профобласти до 2000 элементов
params_prof_only <- expand_grid(
  page = 0:19,
  prof_id = prof_id[prof_id %in% c('17', '38', '153', '171')]
)

### Профобласти более 2000 элементов, только Москва
params_area_exp <- expand_grid(
  page = 0:19,
  prof_id = prof_id[!prof_id %in% c('17', '38', '153', '171')],
  area_id = areas_id[areas_id == 1],
  exp_id = exp_id
)

### Профобласти более 2000 элементов, остальные регионы
params_area <- expand_grid(
  page = 0:19,
  prof_id = prof_id[!prof_id %in% c('17', '38', '153', '171')],
  area_id = areas_id[areas_id != 1]
)


df_prof_only <- params_prof_only %>%
  future_pmap_dfr(get_vacancies_result)


### Запуск
plan(multisession) #параллельные вычисления

start.time <- Sys.time() #время начала 

### 1. Профобласти до 2000 элементов
df_prof_only <- params_prof_only %>%
  future_pmap_dfr(get_vacancies_result)

### 2. Профобласти более 2000 элементов, только Москва
df_area_exp <- params_area_exp %>%
  future_pmap_dfr(get_vacancies_result)

### 3. Профобласти более 2000 элементов, остальные регионы
df_area <- params_area %>%
  future_pmap_dfr(get_vacancies_result)

#Считаем сколько ушло времени
end.time <- Sys.time()
time.taken <- end.time - start.time
print(time.taken)


### Собираем всё в итоговый датасет
vacancies_df <- rbind(df_prof_only, df_area_exp, df_area)



# Записываем вакансии в базу данных #####
## Устанавливаем коннект с базой
connection <- dbConnect(RPostgres::Postgres(), 
                        dbname = 'h0h1_about_hr_analytics', #имя базы
                        host = 'localhost', 
                        port = 5432,
                        user = 'postgres', 
                        password = 'XXXX' #пароль от вашей базы
                        )


## По умолчанию все поля текстовые. Исправляем: числа, как числа и даты, как даты
vacancies_df$id <- as.integer(vacancies_df$id)
vacancies_df$area_id <- as.integer(vacancies_df$area_id)
vacancies_df$professional_roles_id <- as.integer(vacancies_df$professional_roles_id)
vacancies_df$employer_id  <- as.integer(vacancies_df$employer_id)
vacancies_df$created_at <- as.Date(vacancies_df$created_at)
vacancies_df$published_at  <- as.Date(vacancies_df$published_at)


## Записываем таблицу
dbWriteTable(connection, "vacancies_hr", vacancies_df, overwrite = TRUE)

## Индексируем основные поля для повышения скорости работы дашборда
dbExecute(connection, "CREATE INDEX field_indexes ON vacancies_hr (published_at, name, area_name,
                                                                   professional_roles_name,
                                                                   employer_name);")

## Разрываем соединение
dbDisconnect(connection)

File: HR vacancies (hh.ru)/hr_vacancies.py
#!/usr/bin/env python
# coding: utf-8

# # Библиотеки


import requests #для работы с HTTP
import pandas as pd #анализ данных с Пандас
from time import sleep, time #для работы со временем
from concurrent.futures import ThreadPoolExecutor #параллельные вычисления
import itertools #набор инструментов для итераций
from sqlalchemy import create_engine #для записи результатов в БД



# # Основной код

# ## Cоздаем справочники


## Профессиональные области HR
prof_id = [17, 38, 69, 117, 118, 153, 171]

## Регионы
areas_url = "https://api.hh.ru/areas"
areas = requests.get(areas_url).json()
areas_id = [int(area['id']) for area in areas[0]['areas']]

## Опыт работы
exp_id = ['noExperience', 'between1And3', 'between3And6', 'moreThan6']


# ## Определяем функции




#Функция для сбора вакансий в промежуточные датафреймы

def get_vacancies_inter(vacancies):
    
    return pd.DataFrame({
        
        'id': [item['id'] for item in vacancies['items']], #id вакансии
        
        'name': [item['name'] for item in vacancies['items']], #название вакансии
        
        'area_id': [item['area']['id'] for item in vacancies['items']], #id региона
        'area_name': [item['area']['name'] for item in vacancies['items']], #название города
        
        'professional_roles_id': [item['professional_roles'][0]['id'] if item.get('professional_roles') and len(item['professional_roles']) > 0 else None for item in vacancies['items']], #извлекакам id проф. роли
        'professional_roles_name': [item['professional_roles'][0]['name'] if item.get('professional_roles') and len(item['professional_roles']) > 0 else None for item in vacancies['items']], #извлекакам название проф. роли.
        
        'employer_id': [item['employer']['id'] if 'id' in item['employer'] else None for item in vacancies['items']], #id работодателя
        'employer_name': [item['employer']['name'] for item in vacancies['items']], #название компании работодателя
        'snippet_requirement': [item['snippet']['requirement'] if item['snippet'] else None for item in vacancies['items']], #требования к кандидату
        'snippet_responsibility': [item['snippet']['responsibility'] if item['snippet'] else None for item in vacancies['items']], #описание обязанностей
        
        'experience': [item['experience']['name'] for item in vacancies['items']], #требуемый опыт
        'employment': [item['employment']['name'] for item in vacancies['items']],  #тип занятости
        
        'salary_from': [item['salary']['from'] if item.get('salary') and item['salary'].get('from') else None for item in vacancies['items']], #нижняя граница оплаты труда,
        'salary_to': [item['salary']['to'] if item.get('salary') and item['salary'].get('to') else None for item in vacancies['items']], #верхняя граница оплаты труда
        'salary_currency': [item['salary']['currency'] if item.get('salary') and item['salary'].get('currency') else None for item in vacancies['items']], #валюта
        'salary_gross': [item['salary']['gross'] if item.get('salary') and item['salary'].get('gross') else None for item in vacancies['items']], #признак того, что зарплата указана до вычета налогов.
        
        'created_at': [item['created_at'] for item in vacancies['items']], #дата создания вакансии
        'published_at': [item['published_at'] for item in vacancies['items']], #дата публикации вакансии
        
        'url': [item['alternate_url'] for item in vacancies['items']], #ссылка на вакансию на сайте hh.ru
    })





#Итоговая функция по сбору вакансий в один датафрейм
def get_vacancies_result(args):
    
    page = args[0]
    prof_id = args[1]
    area_id = args[2] if len(args) > 2 else None
    exp_id = args[3] if len(args) > 3 else None
    
    vacancies_url = 'https://api.hh.ru/vacancies' #путь к вакансиям API hh.ru
    access_token = 'XXXX' #OAuth токен
    
    headers = {'Authorization': 'Bearer ' + access_token}
    
    #Реализуем логику, отраженную на блок-схеме
    if prof_id in [17, 38, 153, 171]:
        query_list = {'professional_role': prof_id, 'per_page': 100, 'page': page}
    elif area_id == 1:
        query_list = {'professional_role': prof_id, 'area': area_id, 'experience': exp_id, 'per_page': 100, 'page': page}
    else:
        query_list = {'professional_role': prof_id, 'area': area_id, 'per_page': 100, 'page': page}
    
    #Парсим результат по заданному условию, используем для аутентификации OAuth токен
    response = requests.get(vacancies_url, headers=headers, params=query_list)
    
    #Извлекаем данные из JSON
    vacancies = response.json()
    
    #Чтобы избежать ошибки, на тот случай, когда страниц менее 20, пишем такую проверку
    if len(vacancies['items']) == 0:
        return None
    
    #Соединям промежуточные датафреймы в один итоговый
    #Делаем это с интревалом в 0.5 сек, чтобы не получить бан со стороны API
    vacancies_df_inter = get_vacancies_inter(vacancies)
    sleep(0.5)
    
    #Возвращаем результат
    return vacancies_df_inter


# ## Собираем вакансии




## Определяем сетки параметров

### Профобласти до 2000 элементов
params_prof_only = list(itertools.product(range(20), [id for id in prof_id if id in [17, 38, 153, 171]]))

### Профобласти более 2000 элементов, только Москва
params_area_exp = list(itertools.product(range(20), [id for id in prof_id if id not in [17, 38, 153, 171]], [id for id in areas_id if id == 1], exp_id))

### Профобласти более 2000 элементов, остальные регионы
params_area = list(itertools.product(range(20), [id for id in prof_id if id not in [17, 38, 153, 171]], [id for id in areas_id if id != 1]))





### Запуск

start_time = time() #время начала 

with ThreadPoolExecutor() as executor: #параллельные вычисления
    
    ### 1. Профобласти до 2000 элементов
    df_prof_only_results = list(executor.map(get_vacancies_result, params_prof_only))
    df_prof_only_results = [df for df in df_prof_only_results if df is not None]
    df_prof_only = pd.concat(df_prof_only_results, ignore_index=True)
    
    ### 2. Профобласти более 2000 элементов, только Москва
    df_area_exp_results = list(executor.map(get_vacancies_result, params_area_exp))
    df_area_exp_results = [df for df in df_area_exp_results if df is not None]
    df_area_exp = pd.concat(df_area_exp_results, ignore_index=True)
    
    ### 3. Профобласти более 2000 элементов, остальные регионы
    df_area_results = list(executor.map(get_vacancies_result, params_area))
    df_area_results = [df for df in df_area_results if df is not None]
    df_area = pd.concat(df_area_results, ignore_index=True)

#Считаем сколько ушло времени
end_time = time()
time_taken = end_time - start_time

print(time_taken)





### Собираем всё в итоговый датасет
vacancies_df = pd.concat([df_prof_only, df_area_exp, df_area], ignore_index=True)


# ## Записываем вакансии в базу данных




## Переменные для обращения к базе
username = 'postgres'
password = 'XXXX'
host = 'localhost'
port = '5432'
database = 'h0h1_about_hr_analytics'





## По умолчанию все поля object. Исправляем: числа как числа и даты как даты
vacancies_df['id'] = vacancies_df['id'].astype('int64')
vacancies_df['area_id'] = vacancies_df['area_id'].astype('int64')
vacancies_df['professional_roles_id'] = vacancies_df['professional_roles_id'].astype('int64')
vacancies_df['employer_id'] = vacancies_df['employer_id'].fillna(-1).astype('int64')
vacancies_df['salary_from'] = vacancies_df['salary_from'].replace('None', -1).astype('float64')
vacancies_df['salary_to'] = vacancies_df['salary_to'].replace('None', -1).astype('float64')

vacancies_df['salary_gross'] = vacancies_df['salary_gross'].astype(bool)

vacancies_df['created_at'] = pd.to_datetime(vacancies_df['created_at']).dt.date
vacancies_df['published_at'] = pd.to_datetime(vacancies_df['published_at']).dt.date





## Записываем таблицу
engine = create_engine(f'postgresql://{username}:{password}@{host}:{port}/{database}')

vacancies_df.to_sql('vacancies_hr_test', engine, if_exists='replace')





## Индексируем основные поля для повышения скорости работы дашборда
with engine.connect() as connection:
    connection.execute("""
        CREATE INDEX field_indexes_test 
        ON vacancies_hr_test 
        (published_at, name, area_name, professional_roles_name, employer_name);
    """)


